packages/SwingSet/src/devices/vat-admin/device-vat-admin.js

```diff
@@ -214,6 +217,22 @@ export function buildDevice(tools, endowments) {
           return returnFromInvoke(vatID);
         }
 
+        // D(devices.vatAdmin).upgradeVat(bundleID, vatParameters) > upgradeID
+        if (method === 'upgradeVat') {
+          // see #4381, we really need to accept slots in vatParameters
+          // so we can pass contract bundlecaps to upgraded ZCF
+          const { body, slots } = argsCapdata;
+          // this has all the same problems described in terminateWithFailure
+          assert.equal(slots.length, 0, 'upgradeVat() cannot handle refs yet');
+          const args = JSON.parse(body);
+          assert(Array.isArray(args), 'upgradeVat() args array');
+          const [bundleID, vatParameters] = args;
+          assert.typeof(bundleID, 'string', `upgradeVat() bundleID`);
+          const vpCapdata = { body: JSON.stringify(vatParameters), slots };
```

Comment:
Probably should assert that `args` is not only an array but specifically an array of length 2, else the reuse of `slots` on line 231 could pull in other (weird) slot entries that belong to something other than `vatParameters` once `slots` are usable there.  I don't think such extra stuff would introduce a vulnerability or other problems, but it _would_ be a little messy and why take the trouble to analyze such a case when you can just exclude it.